### PR TITLE
basic lock/unlock logic

### PR DESCRIPTION
**Pull request by linuxwolf**

### CODE REVIEW

**Comment by devinreams** [comment]
Sorry for the quick fly-by, but I had two questions for this PR:

1. @linuxwolf are you ready for @jimporter's review? Or do you want to explicitly ask after more work is in place?

2. @m8ttyB @linuxwolf did you two discuss getting Travis set up and configured for this repo too? I'm curious to see how/learn from the difference in config from the extension repo.

**Comment by linuxwolf** [comment]
@devinreams:

1 . Yes, this is ready for review.  Please do so @jimporter 
2. No we haven't talked about it yet.  It is enabled for this repo, waiting on a .travis.yml

File: lib/datastore.js [hunk 1]
```diff
@@ -5,7 +5,52 @@
...
 
 const UUID = require("uuid"),
-      instance = require("./util/instance");
+      errors = require("./util/errors"),
+      instance = require("./util/instance"),
+      jose = require("node-jose");
+
+const DEFAULT_PROMPTS = {
+  unlock: async (request) => {
+    if (1 < request.attempts) {
```
Comment (jimporter) [inline lib/datastore.js]: Nit: `if (request.attempts > 1)` would probably be clearer, but I don't know if we have a specific rule about that.

File: lib/datastore.js [hunk 2]
```diff
@@ -15,44 +60,167 @@ class DataStore {
...
-    inst.items = new Map();
-    inst.locked = true;
+    cfg = cfg || {};
+
+    let self = instance.stage(this);
+    self.items = new Map();
+
+    // TESTING ONLY: accept an (encrypted) empty item keys map
+    self.keystore = cfg.keys;
+    self.prompts = Object.assign({}, DEFAULT_PROMPTS, cfg.prompts);
```
Comment (jimporter) [inline lib/datastore.js]: It's up to you, but we could use the object spread operator here: `{...DEFAULT_PROMPTS, ...cfg.prompts}`. I've been using that in the desktop extension.

File: lib/datastore.js [hunk 3]
```diff
@@ -15,44 +60,167 @@ class DataStore {
...
+   *    request.error;    // an Error describing a previous failure, if any
+   *    request.attempts; // the Number of attempts so far (including this one)
+   *    // returns a Promise containing the unlock secret (as a string)
+   * }
+   * ```
+   *
+   * @type {Object}
+   * @property {Function} unlock The prompt function called to handle {unlock}.
+   */
+  get prompts() { return Object.assign({}, instance.get(this).prompts); }
```
Comment (jimporter) [inline lib/datastore.js]: Ditto here with the object spread operator.

File: lib/datastore.js [hunk 4]
```diff
@@ -15,44 +60,167 @@ class DataStore {
...
+  set prompts(p) {
+    let self = instance.get(this);
+    self.prompts = Object.assign({}, DEFAULT_PROMPTS, p || {});
```
Comment (jimporter) [inline lib/datastore.js]: And here with the object spread operator.

File: lib/util/errors.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+["ABORTED", "UNLOCK_ERROR", "SYNC_ERROR", "GENERIC_ERROR"].forEach(m => {
+  let s = Symbol(m);
+  Reasons[m] = s;
+  Messages[s] = m;
+});
+
+class DataStoreError extends Error {
+  constructor(message, reason) {
+    if ("symbol" === typeof message) {
```
Comment (jimporter) [inline lib/util/errors.js]: `if (typeof message === "symbol")`?

File: lib/util/errors.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      reason = Reasons.GENERIC_ERROR;
+    }
+    if (!message) {
+      message = Messages[reason];
+    }
+
+    super(message);
+    this.name = this.constructor.name;
+    this.reason = reason;
+    if ("function" === Error.captureStackTrace) {
```
Comment (jimporter) [inline lib/util/errors.js]: `if (Error.captureStackTrack === "function")`?

**Comment by linuxwolf** [comment]
I'll add an issue to consider object spread, but for now just merge this.

**Comment by linuxwolf** [comment]
closes #2

**Pull request merged**